core/solver/multigrid.cpp

```diff
@@ -581,61 +531,126 @@ void Multigrid::generate()
+                dense_x->scale(dense_beta);
+                dense_x->add_scaled(dense_alpha, x_clone.get());
+            },
+            alpha, b, beta, x);
+    };
+    auto first_mg_level = this->get_mg_level_list().front();
+    run<gko::multigrid::EnableMultigridLevel, float, double,
+        std::complex<float>, std::complex<double>>(first_mg_level, lambda,
+                                                   alpha, b, beta, x);
+}
+
+
+template <typename VectorType>
+void Multigrid::apply_dense_impl(const VectorType* b, VectorType* x,
+                                 initial_guess_mode guess) const
+{
+    using ws = workspace_traits<Multigrid>;
+    this->setup_workspace();
+    this->create_state();
+    if (cache_.state->nrhs != b->get_size()[1]) {
+        cache_.state->generate(this->get_system_matrix().get(), this,
+                               b->get_size()[1]);
+    }
+    auto lambda = [&, this](auto mg_level, auto b, auto x) {
         using value_type = typename std::decay_t<
             gko::detail::pointee<decltype(mg_level)>>::value_type;
         auto exec = this->get_executor();
-        auto neg_one_op =
+        static auto neg_one_op =
             initialize<matrix::Dense<value_type>>({-one<value_type>()}, exec);
-        auto one_op =
+        static auto one_op =
             initialize<matrix::Dense<value_type>>({one<value_type>()}, exec);
         constexpr uint8 RelativeStoppingId{1};
-        array<stopping_status> stop_status(exec, b->get_size()[1]);
+        auto& stop_status =
+            this->template create_workspace_array<stopping_status>(
+                ws::stop, b->get_size()[1]);
         bool one_changed{};
-        auto state =
-            MultigridState(system_matrix_.get(), this, b->get_size()[1]);
         exec->run(multigrid::make_initialize(&stop_status));
         // compute the residual at the r_list(0);
-        auto r = state.r_list.at(0);
-        r->copy_from(b);
-        if (parameters_.zero_guess) {
-            using matrix::Dense;
-            using std::complex;
-            if (auto dense = dynamic_cast<Dense<float>*>(x)) {
-                dense->fill(zero<float>());
-            } else if (auto dense = dynamic_cast<Dense<double>*>(x)) {
-                dense->fill(zero<double>());
-            } else if (auto dense = dynamic_cast<Dense<complex<float>>*>(x)) {
-                dense->fill(zero<complex<float>>());
-            } else if (auto dense = dynamic_cast<Dense<complex<double>>*>(x)) {
-                dense->fill(zero<complex<double>>());
-            } else {
-                GKO_NOT_SUPPORTED(x);
-            }
-        }
-        system_matrix_->apply(lend(neg_one_op), x, lend(one_op), r.get());
-        auto stop_criterion = stop_criterion_factory_->generate(
-            system_matrix_,
+        // auto r = state->r_list.at(0);
+        // r->copy_from(b)
+        // system_matrix->apply(lend(neg_one_op), x, lend(one_op), r.get());
```

Comment:
nit: commented code (I'm assuming this is from debugging which is not done yet).
Applies to the next ~20 lines as well.